### ISSUE TITLE
Invalid ngOnDestroy invocation on multiple provision

### ISSUE THREAD

**Issue by Azbesciak** [angular/angular#35231]
# 🐞 bug report
### Affected Package
core? 9.0.0. Would like to know
### Is this a regression?
It occured after 

 Yes, the previous version in which this bug was not present was: ....

### Description
Not sure whether it is connected with https://github.com/angular/angular/issues/35167 or not, but following screen shows an error:
![image](https://user-images.githubusercontent.com/17334466/74039843-da1c6b00-49c2-11ea-8570-eebeb81f6482.png)
in summary, those services in array - all of them implements some interface and are provided with the same token, under provision with `multiple: true`. As you see, `ngOnDestroy` is invoked on that array, but this `this`... eh, I don't think that this array extends `Subscriber`. I have also noticed that this is not always the case.

## 🔥 Exception or Error
<pre><code>
ERROR Error: Uncaught (in promise): TypeError: this.clearSubscriptions is not a function
TypeError: this.clearSubscriptions is not a function
    at Array.ngOnDestroy (subscriber.ts:28)
    at executeOnDestroys (core.js:14024)
    at cleanUpView (core.js:13931)
    at destroyViewTree (core.js:13705)
    at destroyLView (core.js:13882)
    at RootViewRef.destroy (core.js:14737)
    at ComponentRef$1.destroy (core.js:33528)
    at RouterOutlet.deactivate (router.js:8912)
    at ActivateRoutes.deactivateRouteAndOutlet (router.js:3963)
    at ActivateRoutes.deactivateRouteAndItsChildren (router.js:3921)
    at resolvePromise (zone-evergreen.js:793)
    at resolvePromise (zone-evergreen.js:752)
    at zone-evergreen.js:854
    at ZoneDelegate.invokeTask (zone-evergreen.js:400)
    at Object.onInvokeTask (core.js:40744)
    at ZoneDelegate.invokeTask (zone-evergreen.js:399)
    at Zone.runTask (zone-evergreen.js:168)
    at drainMicroTaskQueue (zone-evergreen.js:570)
    at invokeTask (zone-evergreen.js:485)
    at ZoneTask.invoke (zone-evergreen.js:470)
</code></pre>

## 🌍  Your Environment

**Angular Version:**
<pre><code>
Angular CLI: 9.0.1
Node: 12.13.1
OS: win32 x64

Angular: 9.0.0
... animations, cdk, common, compiler, compiler-cli, core
... elements, forms, language-service, material
... platform-browser, platform-browser-dynamic, router
Ivy Workspace: Yes

Package                            Version
------------------------------------------------------------
@angular-devkit/architect          0.900.1
@angular-devkit/build-angular      0.900.1
@angular-devkit/build-ng-packagr   0.900.1
@angular-devkit/build-optimizer    0.900.1
@angular-devkit/build-webpack      0.900.1
@angular-devkit/core               8.3.23
@angular-devkit/schematics         8.3.23
@angular/cli                       9.0.1
@ngtools/webpack                   9.0.1
@schematics/angular                8.3.23
@schematics/update                 0.900.1
ng-packagr                         9.0.0
rxjs                               6.5.4
typescript                         3.7.5
webpack                            4.41.2
</code></pre>

**Comment by pkozlowski-opensource** [comment]
This one sounds very much like #35167 ...

**Comment by crisbeto** [comment]
I could reproduce the issue while working on #35167 and I'm not sure that it's the same problem. @Azbesciak are you sure that this even worked with ViewEngine? Running my test against it shows that the `ngOnDestroy` was never called in the first place. Here's the test:

```
it('', () => {
  const destroyContexts: (DestroyService | undefined)[] = [];
  const services = new InjectionToken<any>('t');

  @Injectable()
  class DestroyService {
    ngOnDestroy() { destroyContexts.push(this); }
  }

  @Injectable()
  class OtherDestroyService {
    ngOnDestroy() { destroyContexts.push(this); }
  }

  @Component({
    template: '<div></div>',
    providers: [
      {provide: services, useClass: DestroyService, multi: true},
      {provide: services, useClass: OtherDestroyService, multi: true}
    ]
  })
  class App {
    constructor(@Inject(services) s: any) {}
  }

  TestBed.configureTestingModule({declarations: [App]});
  const fixture = TestBed.createComponent(App);
  fixture.detectChanges();
  fixture.destroy();

  console.log(destroyContexts);
});
```

In Ivy this logs out `[[DestroyService, OtherDestroyService], App]` (obviously wrong).
In ViewEngine the `destroyContexts` is empty. I'm unsure whether this was working as expected or it
was undefined ViewEngine behavior.

**Comment by crisbeto** [comment]
Removing the `has PR` label, because we decided to decouple the fix for this from the fix for #35167. I've got a fix for this issue in a separate branch, but I need to set up some benchmarks first.

**Comment by Azbesciak** [comment]
Is this going to be merged?

**Comment by crisbeto** [comment]
It's going to be fixed, but we decided to decouple the fix from #35249.
